### PR TITLE
httpcli: switch outbound request logger to FIFOList

### PR DESCRIPTION
**Pull request by keegancsmith**
The use of AllKeys had a hidden cost of scanning _every_ key in redis,
not just those associated with outbound requests. So this commit ports
the outbound-requests logger to use a FIFOList, the same datastructure
used by our slow request logger.

This has the additional benefit of reducing our API surface in rcache,
since this was the only user of AllKeys and the Multi* functions.

We could likely do this better since we do things like lookup all logged
requests each time we want to read. But given the value is small (500
items) and it is only done on read (not log time), this seems like a
fine tradeoff versus doing a bigger change.

Test Plan: go test

plz-review-url: https://plz.review/review/17327

### CODE REVIEW

File: internal/httpcli/redis_logger_middleware.go [hunk 1]
```diff
@@ -18,7 +18,11 @@ import (
 	"github.com/sourcegraph/sourcegraph/lib/errors"
 )
 
-var outboundRequestsRedisCache = rcache.NewWithTTL("outbound-requests", 604800)
+// outboundRequestsRedisFIFOListDefaultSize sets a default value for the FIFO list.
+const outboundRequestsRedisFIFOListDefaultSize = 50
+
+// outboundRequestsRedisFIFOList is a FIFO redis cache to store the requests.
+var outboundRequestsRedisFIFOList = rcache.NewFIFOList("outbound-requests", outboundRequestsRedisFIFOListDefaultSize)
```
Comment (keegancsmith) [inline internal/httpcli/redis_logger_middleware.go:25]: should be fine to use the same key name, but maybe I should use a different one just to be safe? Previously all keys actually would of been `outbound-requests:item-id`, but nothing would of actually been stored at `outbound-requests`.

**Review by keegancsmith — COMMENTED** [review]
FYI I still wanna do some manual testing

File: internal/httpcli/redis_logger_middleware.go (hunk 1 above)
Comment (mrnugget) [reply internal/httpcli/redis_logger_middleware.go:25]: Seems fine if the previous key was just the "prefix"

File: internal/httpcli/redis_logger_middleware.go [hunk 2]
```diff
@@ -106,103 +113,82 @@ func redisLoggerMiddleware() Middleware {
...
 
-// getAllValuesAfter returns all items after the given key, in ascending order, trimmed to maximum {limit} items.
-func getAllValuesAfter(ctx context.Context, c *rcache.Cache, after string, limit int) ([][]byte, error) {
-	all, err := c.ListKeys(ctx)
-	if err != nil {
-		return nil, err
-	}
+// getOutboundRequestLogItems returns all items where pred returns true,
+// sorted by ID ascending.
+func getOutboundRequestLogItems(ctx context.Context, pred func(*types.OutboundRequestLogItem) bool) ([]*types.OutboundRequestLogItem, error) {
```
Comment (mrnugget) [inline internal/httpcli/redis_logger_middleware.go:168]: Do we still have a test for this function?

File: internal/httpcli/redis_logger_middleware.go [hunk 3]
```diff
@@ -106,103 +113,82 @@ func redisLoggerMiddleware() Middleware {
...
+	// We fetch all values from redis, then just return those in after. Given
```
Comment (mrnugget) [inline internal/httpcli/redis_logger_middleware.go:169]: What is "in after" referring to?

**Review by mrnugget — APPROVED** [review]

File: internal/httpcli/redis_logger_middleware.go (hunk 3 above)
Comment (keegancsmith) [reply internal/httpcli/redis_logger_middleware.go:169]: sorry after was a leftover from older code. Will push a commit referring to pred instead.

File: internal/httpcli/redis_logger_middleware.go (hunk 2 above)
Comment (keegancsmith) [reply internal/httpcli/redis_logger_middleware.go:168]: it is tested via the tests which exercise `GetOutboundRequestLogItems`. However, I spent some time increasing test coverage of `GetOutboundRequestLogItems` and added a test for `GetOutboundRequestLogItem`. See the pushed commits.

**Comment by keegancsmith** [comment]
Alright did a bunch of manual testing plus added another test to ensure after still works.

**Pull request merged**